### PR TITLE
Add Ability for User to Customize Email Notification Frequency

### PR DESCRIPTION
**Pull request by gary-kim**
UI:
Screenshot from 2019-08-16 23-14-14

---

This commit adds the backend logic for
allowing users to enable or disable email
notifications. The implementation ensures
that only issue notification emails get disabled
and important emails are still sent regardless
of the setting.

The UI to toggle this setting has not yet been
implemented.

---

This commit completes the functionality to allow
users to disable their own email notifications.

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -719,6 +720,18 @@ func (u *User) IsMailable() bool {
 	return u.IsActive
 }
 
+// EnabledEmailNotifications checks if the user has
+// enabled receiving notifications by email
+func (u *User) EnabledEmailNotifications() bool {
```
Comment (lafriks) [inline models/user.go]: Why not just add this condition to IsMailable?

Comment (gary-kim) [reply models/user.go]: I was thinking that maybe if other email notifications are added, we'd want a distinction between whether the user chose not to receive emails or whether they are not active. So for example, security related emails get sent regardless of the user's email notification choice but doesn't get sent if the user is not active.

**Comment by lunny** [comment]
I would like the third option `send mail only mentioned me`.

**Comment by gary-kim** [comment]
> I would like the third option `send mail only mentioned me`.

That is a good idea. I'll add it here for now but I think in the long run, that should be an option that is repository specific.

**Comment by gary-kim** [comment]
@lunny Could you take another look?

File: models/migrations/v93.go [hunk 2]
```diff
@@ -9,7 +9,7 @@ import "github.com/go-xorm/xorm"
 func addEmailNotificationEnabledToUser(x *xorm.Engine) error {
 	// Issue see models/user.go
 	type User struct {
-		EmailNotificationsEnabled bool `xorm:"NOT NULL DEFAULT true"`
+		EmailNotificationsEnabled bool `xorm:"DEFAULT true"`
```
Comment (lafriks) [inline models/migrations/v93.go]: Why remove `NOT NULL`?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/migrations/v93.go (hunk 2 above)
Comment (gary-kim) [reply models/migrations/v93.go]: Sorry, that is from back when that was still a bool. I'll put it back on.

**Comment by guillep2k** [comment]
Could you add some tests for this?

**Comment by gary-kim** [comment]
@guillep2k Done.

**Comment by guillep2k** [comment]
> 
> 
> @guillep2k Done.

Sorry, I was expecting something for checking if the actions (e-mails) are generated in the correct cases, i.e., create an issue, comment on an existing issue  and check which e-mails were generated and which weren't. Like this:

https://github.com/go-gitea/gitea/blob/ca6fb004ac50fc924861112403895d637c6a2d1d/models/repo_watch_test.go#L71-L87

Checking the e-mail generation is useful for the future, as many changes in the code made by others could inadvertently break this functionality. A test will help on that case.

**Comment by gary-kim** [comment]
I'm not sure how that would work for the email system in its current state. The function that has to be run to check will send emails on its own. Maybe it could be done as part of integrations but I'm a bit short on time at the moment.

**Comment by FalconWu2017** [comment]
Can change the selected option for every user by Adminatrator?

**Comment by gary-kim** [comment]
> Can change the selected option for every user by Adminatrator?

Not currently, I'll add a default option that admins can choose.

**Comment by gary-kim** [comment]
@lunny @lafriks 
Could you take a look?

File: models/issue_mail.go [hunk 3]
```diff
@@ -70,17 +70,17 @@ func mailIssueCommentToParticipants(e Engine, issue *Issue, doer *User, content
 		if err != nil {
 			return fmt.Errorf("GetUserByID [%d]: %v", watchers[i].UserID, err)
 		}
-		if to.IsOrganization() {
+		if to.IsOrganization() || to.EmailNotifications() == EmailNotificationsEnabled {
```
Comment (lafriks) [inline models/issue_mail.go]: Is this condition really right? Shouldn't it be:
```suggestion
		if to.IsOrganization() || to.EmailNotifications() != EmailNotificationsEnabled {
```

File: models/issue_mail.go [hunk 4]
```diff
@@ -70,17 +70,17 @@ func mailIssueCommentToParticipants(e Engine, issue *Issue, doer *User, content
...
 		tos = append(tos, to.Email)
 		names = append(names, to.Name)
 	}
 	for i := range participants {
-		if participants[i].ID == doer.ID {
-			continue
-		} else if com.IsSliceContainsStr(names, participants[i].Name) {
+		if participants[i].ID == doer.ID ||
+			com.IsSliceContainsStr(names, participants[i].Name) ||
+			participants[i].EmailNotifications() == EmailNotificationsEnabled {
```
Comment (lafriks) [inline models/issue_mail.go]: same here

File: models/issue_mail.go (hunk 3 above)
Comment (gary-kim) [reply models/issue_mail.go]: Whoops, not sure how I missed that. I've changed it now.

File: models/user.go [hunk 5]
```diff
@@ -719,6 +727,17 @@ func (u *User) IsMailable() bool {
...
 }
 
+// EmailNotifications returns the User's email notification preference
+func (u *User) EmailNotifications() string {
+	return u.EmailNotificationsPreference
+}
+
+// SetEmailNotifications sets the user's email notification preference
+func (u *User) SetEmailNotifications(set string) {
+	u.EmailNotificationsPreference = set
```
Comment (lunny) [inline models/user.go:737]: This is not a good usage, we should update user by the specific columns.

**Review by lunny — CHANGES_REQUESTED** [review]
And how `onmention` work? It seems there is no related code changes.

**Comment by gary-kim** [comment]
> And how `onmention` work? It seems there is no related code changes.

The check for watch and participants is "do not send if notifications are not enabled"  so `onmention` will not be sent, and the check for mentions are "send emails if not disabled" so `onmention` will be sent.

File: models/user.go (hunk 5 above)
Comment (gary-kim) [reply models/user.go:737]: Okay, updating the specific column now.

File: models/user.go [hunk 6]
```diff
@@ -719,6 +727,20 @@ func (u *User) IsMailable() bool {
...
+	err := UpdateUserCols(u, "email_notifications_preference")
```
Comment (lafriks) [inline models/user.go]:
```suggestion
	if err := UpdateUserCols(u, "email_notifications_preference"); err != nil {
```

File: models/migrations/v93.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import "github.com/go-xorm/xorm"
+
+func addEmailNotificationEnabledToUser(x *xorm.Engine) error {
+	// Issue see models/user.go
+	type User struct {
+		EmailNotificationsPreference string `xorm:"NOT NULL DEFAULT 'enabled'"`
```
Comment (lunny) [inline models/migrations/v93.go]: default string is `VARCHAR(255)`, I think it's unnecessary to use 255 bytes. Maybe 10 or 20 is enough.

**Review by lunny — APPROVED** [review]

**Review by guillep2k — CHANGES_REQUESTED** [review]

**Comment by guillep2k** [comment]
As for the build error, please check the conversation here:
https://github.com/go-gitea/gitea/pull/8021#issuecomment-525940166

**Review by guillep2k — APPROVED** [review]

File: models/migrations/v93.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import "github.com/go-xorm/xorm"
+
+func addEmailNotificationEnabledToUser(x *xorm.Engine) error {
+	// Issue see models/user.go
```
Comment (lafriks) [inline models/migrations/v93.go]:
```suggestion
	// User see models/user.go
```

**Comment by lafriks** [comment]
nit in migration comment but otherwise LGTM

**Review by lafriks — APPROVED** [review]

**Review by guillep2k — APPROVED** [review]

**Comment by lafriks** [comment]
`modules.txt` needs to updated with value `golang.org/x/crypto v0.0.0-20190829043050-9756ffdc2472` for `golang.org/x/crypto` line

**Comment by mario-neuhold** [comment]
It seems that `onmention` does not trigger sending a notifications for when a user is assigned to an issue. Does it make sense to include this scenario?

**Comment by guillep2k** [comment]
@mario-neuhold AFAICT assigments should send an e-mail notification to the assigned user (if they have notifications enabled) as of version 1.11.1. If you're having problem with that, could you please open an issue and provide details?

**Pull request merged**